pkg/fatfs/fatfs_diskio/mtd/mtd_diskio.c

```diff
@@ -130,24 +130,24 @@ DRESULT disk_read(BYTE pdrv, BYTE *buff, DWORD sector, UINT count)
  */
 DRESULT disk_write(BYTE pdrv, const BYTE *buff, DWORD sector, UINT count)
 {
+    mtd_dev_t *mtd = fatfs_mtd_devs[pdrv];
     DEBUG("disk_write: %d, %lu, %d\n", pdrv, (long unsigned)sector, count);
-    if ((pdrv >= FF_VOLUMES) || (fatfs_mtd_devs[pdrv]->driver == NULL)) {
+    if ((pdrv >= FF_VOLUMES) || (mtd->driver == NULL)) {
         return RES_PARERR;
     }
 
-    /* erase memory before writing to it */
-    int res = mtd_erase_sector(fatfs_mtd_devs[pdrv], sector, count);
-
-    if (res != 0) {
-        return RES_ERROR; /* erase failed! */
+    if (mtd_sector_needs_erase(mtd)) {
```

Comment:
I wonder if this should really be part of the application or not better done implicitly by the driver in `mtd_write_page_raw`.

Reply:
The idea is that `mtd_write_page_raw()` should really just do write.
Maybe a better option would be to introduce `mtd_write_sector()` for when we want to write whole sectors - MTD can then deal with erasing them or not.

Reply:
> The idea is that `mtd_write_page_raw()` should really just do write.

I see.

> Maybe a better option would be to introduce mtd_write_sector() for when we want to write whole sectors - MTD can then deal with erasing them or not.

It doesn't matter for SD Cards since `pages_per_sector` is 1 and the `page_size` corresponds to the `write_size` which is the SDHC block size of 512 bytes.

Reply:
Yes, the idea is more that if we write a whole sector (which is the erase unit) we don't have to perform a read-modify-write cycle since everything will be overwritten / erased.

So we can auto-erase the sector if needed (like `mtd_write_page()` but don't have to care about the previous contents (like `mtd_write_page_raw()`).

Reply:
So you would call `mtd_write_sector()` here instead of `mtd_write_page_raw()` which then takes care whether it is necessary to call `mtd_erase_sector()` before writing the whole sector.